### PR TITLE
vrg: don't watch velero CRs unless CRDs are installed

### PR DESCRIPTION
**Pull request by raghavendra-talur**
Throw error in the reconcile function if velero is not installed and a
kube object protection VRG is created.

Signed-off-by: Raghavendra Talur <raghavendra.talur@gmail.com>

### CODE REVIEW

**Review by rakeshgm — APPROVED** [review]

File: controllers/volumereplicationgroup_controller.go [hunk 1]
```diff
@@ -1512,3 +1518,41 @@ func (r *VolumeReplicationGroupReconciler) addVolsyncOwnsAndWatches(ctrlBuilder
...
+		installedCRD := &apiextensionsv1.CustomResourceDefinition{}
+		if err := r.APIReader.Get(context.TODO(), types.NamespacedName{Name: crd}, installedCRD); err != nil {
+			r.Log.Info("Velero CRD not found", "CRD", crd, "error", err)
+
+			foundCRDS = false
+		}
+	}
+
+	if !foundCRDS {
+		r.Log.Info("Velero CRDs not found; Kubernetes object protection won't work unless Velero/OADP is installed")
```
Comment (nirs) [inline controllers/volumereplicationgroup_controller.go:1545]: Need to unify with the reconcile error message.

**Review by nirs — COMMENTED** [review]
Not clear what is expected the outcome - if some required CRD is missing we will see one log for missing CRD and one log about installing velero/openshift-adp (good). But what will happen when reconciling a VRG with kube object protection? How many errors are expected? hopefully one on the first reconcile?

File: controllers/volumereplicationgroup_controller.go [hunk 2]
```diff
@@ -433,6 +435,11 @@ func (r *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
 		v.instance.Spec.Async, cephFSCSIDriverNameOrDefault(v.ramenConfig),
 		volSyncDestinationCopyMethodOrDefault(v.ramenConfig), adminNamespaceVRG)
 
+	if adminNamespaceVRG && !r.veleroCRsAreWatched {
```
Comment (BenamarMk) [inline controllers/volumereplicationgroup_controller.go:438]: Move this block to line 433. You don't need to create a new VolSync handler if we are going to fail it.

Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go:438]: Done

File: controllers/volumereplicationgroup_controller.go [hunk 3]
```diff
@@ -433,6 +434,11 @@ func (r *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
...
+		return ctrl.Result{}, fmt.Errorf("VRG with kube object protection doesn't work if velero/oadp is not installed. " +
+			"Please install velero/oadp and restart the operator")
```
[the remark this answers is not in the document]
Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go:440]: Changed the error message to include the namespace and name.

The outcome is that the reconcile for *this* vrg will start failing and retried with a ExponentialBackoff, so we will see the error once in 18 minutes eventually.

File: controllers/volumereplicationgroup_controller.go [hunk 4]
```diff
@@ -1512,3 +1518,41 @@ func (r *VolumeReplicationGroupReconciler) addVolsyncOwnsAndWatches(ctrlBuilder
...
+		"backupstoragelocations.velero.io",
+		"restores.velero.io",
+	}
+
+	foundCRDS := true
+
+	for _, crd := range veleroCRDs {
+		installedCRD := &apiextensionsv1.CustomResourceDefinition{}
+		if err := r.APIReader.Get(context.TODO(), types.NamespacedName{Name: crd}, installedCRD); err != nil {
+			r.Log.Info("Velero CRD not found", "CRD", crd, "error", err)
```
[the remark this answers is not in the document]
Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go:1538]: I think it is ok to return even for temporary errors as the reconciler will take care of retries. For 4.16 I don't want to be intelligent about error handling and just log it. May be we can handle specific errors later.

File: controllers/volumereplicationgroup_controller.go [hunk 5]
```diff
@@ -1512,3 +1518,41 @@ func (r *VolumeReplicationGroupReconciler) addVolsyncOwnsAndWatches(ctrlBuilder
...
+	r.Log.Info("Kube object protection enabled; watch kube objects requests")
+
+	// Find if velero CRDs are present in the cluster
+	veleroCRDs := []string{
+		"backups.velero.io",
+		"backupstoragelocations.velero.io",
+		"restores.velero.io",
+	}
+
+	foundCRDS := true
```
[the remark this answers is not in the document]
Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go:1533]: Done

File: controllers/volumereplicationgroup_controller.go (hunk 1 above)
Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go:1545]: The reconcile error message is in addition to this log. I think this message can be a shorter one.

File: controllers/volumereplicationgroup_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nirs) [reply controllers/volumereplicationgroup_controller.go:1539]: Using simple error handling for minimal change sounds good! but lets not lie about the error and use "Cannot get velero CRD" - the error will explain why (e.g. not found).

[the remark this answers is not in the document]
Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go]: Done

File: controllers/volumereplicationgroup_controller.go (hunk 1 above)
Comment (nirs) [reply controllers/volumereplicationgroup_controller.go]: Sure but we need to use the same terms and style in both logs. Currently we have:
- Velero/OADP
- velero/oadp

My suggestion velero/openshift-adp in both.

Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go]: I need to remove this error message as this gets logged for Managed App VRGs too.

Comment (raghavendra-talur) [reply controllers/volumereplicationgroup_controller.go]: I verified, it only logs once when the operator starts. That is ok.

File: controllers/volumereplicationgroup_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/volumereplicationgroup_controller.go]: Returning an error is the right course of action. Perhaps we can add an alert in a future PR.

**Review by BenamarMk — APPROVED** [review]

File: controllers/volumereplicationgroup_controller.go [hunk 6]
```diff
@@ -429,6 +430,12 @@ func (r *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
 	v.ramenConfig = ramenConfig
 	adminNamespaceVRG := vrgInAdminNamespace(v.instance, v.ramenConfig)
 
+	if adminNamespaceVRG && !r.veleroCRsAreWatched {
+		return ctrl.Result{},
+			fmt.Errorf("VRG {%s/%s} with kube object protection doesn't work if velero/oadp is not installed. "+
+				"Please install velero/oadp and restart the operator", v.instance.Namespace, v.instance.Name)
+	}
```
Comment (nirs) [inline controllers/volumereplicationgroup_controller.go:438]: Can you share actual logs showing these errors? I think we will see lot of unwanted and unhelpful repeated messages with this. Each one will have unhelpful trackback.

**Pull request merged**